Add DAL functions in preparation for voicemail admin tool

jrrose opened the pull request:
context: all() - retrieves all contexts
folder: get(name) - retrieves a folder by name
folder: getNameOrDTMF(name, dtmf) - retrieves all folders matching either the name or the dtmf
mailbox: allContext(context) - retrieves all mailboxes matching the context
mailbox: countContext(context) - counts all mailboxes matching the context
message: removeMailbox(mailbox) - removes all messages belonging to a specific mailbox

jrrose committed:
Add dal functions in preparation for admin tool.

jrrose committed:
Remove unfinished commented out test.

samuelg commented on lib/repositories/folder.js:
```diff
@@ -99,6 +99,58 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+     * the name or DTMF provided. Useful for checking for conflicts
+     * prior to adding a new item.
+     */
+    getNameOrDTMF: function(name, dtmf) {
```
get should be reserved for fetching a single instance. I would rename this method findByNameOrDTMF.

samuelg commented on lib/repositories/mailbox.js:
```diff
@@ -82,6 +82,48 @@ function createApi(table, provider, sqlGenerator, dependencies) {
     /**
+     * Return all mailboxes belonging to the specified context
+     */
+    allContext: function(context) {
```
all should be reserved for fetching all instances. I would rename this findByContext.

samuelg commented on lib/repositories/mailbox.js:
```diff
@@ -82,6 +82,48 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+    /**
+     * Returns count of all mailboxes belonging to the specified context
+     */
+    countContext: function(context) {
```
I would rename this countByContext.

samuelg commented on lib/repositories/message.js:
```diff
@@ -394,6 +424,43 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+    /**
+     * Deletes all messages belonging to a specific mailbox
+     */
+    removeMailbox: function remove(mailbox) {
```
I would rename this removeForMailbox or removeByMailbox.

samuelg replied:
The existing remove method returns the message instance that was just deleted so that the caller can update MWI counts via ARI so either decrement the read or unread messages accordingly. We probably need to return an array of message instances deleted so that the caller can do the same.

Or will you be deleting the MWI mailbox counts as part of calling this? Still sounds like it might be a good idea to add this in case the caller wants to do something like this.

jrrose replied:
I hadn't actually thought about MWI at the time, but it seems like the obvious thing to do would be to just set both read and unread to 0 when all the messages are deleted for a mailbox.

samuelg commented:
We definitely need unit tests for these new methods before we can merge this PR.

samuelg commented on lib/repositories/context.js:
```diff
@@ -60,6 +60,32 @@ function createApi(table, provider, sqlGenerator, dependencies) {
     },
 
     /**
+     * Returns an array where the values are context instances.
```
This does not actually return an array. It returns a promise that will resolve to an array. I would recommend adding jshint style metadata to the method comments.

samuelg replied:
Disregard if you like. It seems that's the convention in that file. Should probably be addressed at some point in the future however.

samuelg commented on lib/repositories/mailbox.js:
```diff
@@ -82,6 +82,48 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+
+      function constructor(id) {
+        /*jshint validthis:true*/
+        return this.create(undefined, context, undefined, id);
```
This needs  a comment to explain why we are passing undefined to the create method.

samuelg replied:
Fields is optional, you don't have to pass undefined for that yourself.

samuelg commented on lib/repositories/message.js:
```diff
@@ -132,6 +132,36 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+    /**
+     * Returns a count of all messages for the given mailbox.
+     */
+    countMailbox: function(mailbox) {
```
I would rename this countByMailbox.

samuelg commented on lib/repositories/message.js:
```diff
@@ -132,6 +132,36 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+    /**
+     * Returns a count of all messages for the given folder.
+     */
+    countFolder: function(folder) {
```
I would rename this countByFolder.

samuelg commented:
It feels like the various findBy_ and countBy_ methods could be abstracted out. There's a lot code reuse there. Perhaps there's a way to built a findByProperty and countByProperty. It's something to consider.

jrrose committed:
Add tests, address findings, rename functions.

jrrose commented:
Alright, I've posted a new commit that adds tests and addresses the above findings to at least the degree that I felt confident addressing them. It also addresses the timeout issues we were discussing earlier by upping timeout values by a factor a three.

matt-jordan commented on lib/repositories/context.js:
```diff
@@ -60,6 +60,32 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+
+      function constructor(id) {
+        /*jshint validthis:true*/
+        //Passing undefined so that domain can be populated by db result
```
Space between '//' and the rest of the comment

matt-jordan commented on lib/repositories/message.js:
```diff
@@ -132,6 +132,24 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+     * Returns a count of all messages for the given mailbox.
+     */
+    countByMailbox: function(mailbox) {
+      dependencies.logger.trace('message.countMailbox called');
```
s/countMailboox/countByMailbox

matt-jordan commented on lib/repositories/message.js:
```diff
@@ -394,20 +412,88 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+                  return transaction.commit();
+                })
+                .then(function() {
+                  dependencies.logger.debug('Messages removed');
```
This might be more useful if it logged out which messages are removed.

jrrose committed:
Address comments by matt-jordan on pull request

Also upped timeout for mailbox tests to 200ms from 150ms
after hitting an unlucky timeout failure.

jrrose committed:
Pull out find(where) function from folder dal

This was a function that I added to reduce duplicate code,
but in retrospect it isn't something that should be exposed
to consumers. I've pulled it out of the exposed API and made
it private. Not sure if I did this in the most elegant way.

jrrose committed:
Mess with the folder find function some more.

Made it so that less stuff had to be passed to it by moving it to be
within the scope of createApi.

jrrose commented on lib/repositories/message.js:
```diff
@@ -394,20 +412,89 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+    /**
+     * Deletes all messages belonging to the specific mailbox
+     */
+    removeByMailbox: function remove(mailbox) {
```
So I noticed I duplicated this construct from the normal remove function... is there any meaning to having 
function function_name(args)
like in this example instead of just
function (args)
like what we do elsewhere?

It doesn't seem to have any effect on behavior, but I figure it's something I should probably clean up.

samuelg replied:
Since you name the function removeByMailbox, I would remove the remove name after function. Some people encourage this so that stack traces can show the name of the function where an error occured. In that case however, it should read removeByMailbox: function removeByMailbox(mailbox).

I don't personally like the duplication of names so I don't bother with it.

samuelg replied:
I would remove 'remove' from function remove.

samuelg commented on lib/repositories/folder.js:
```diff
@@ -75,22 +75,30 @@ function createApi(table, provider, sqlGenerator, dependencies) {
-        .toQuery();
-
-      return common.find(query, provider, constructor.bind(this))
+      return find(this)
```
This may reuse some code, but it's less readable than it was before. A call to the find function passing only a this reference under the all method is confusing. I would revert all to what it was before and put the full where clause in the findByNameOrDTMF method. That way is method is self contained and uses a predictable API.

samuelg commented on lib/repositories/mailbox.js:
```diff
@@ -82,6 +82,49 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+      function constructor(id) {
+        /*jshint validthis:true*/
+        //Passing undefined so that number can be populated by db result
+        return this.create(undefined, context, {}, id);
```
The third param can either be an object representing fields or an id. You can just remove that parameter, it's intended to work that way.

samuelg commented on lib/repositories/message.js:
```diff
@@ -394,20 +412,89 @@ function createApi(table, provider, sqlGenerator, dependencies) {
-     var query = table
+    var where;
+
+    if ('mailbox' in fields) {
```
This is way too complicated for a count method. I would revert this method and just call the expected API for calculating a count. Duplication of a consistent API is not necessarily bad; especially if abstracting out the calls ends up less readable than the original format.

samuelg commented on lib/repositories/message.js:
```diff
@@ -394,20 +412,89 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+              var messages = result.rows;
+
+              var x;
+              for (x in messages) {
```
I would duplicate how this is done in other parts of the DAL and use result.reduce to construct the output.

samuelg replied:
Actually result.rows.map(convertFromStorage) should work even better.

samuelg commented on test/context.js:
```diff
@@ -89,6 +90,58 @@ describe('context', function () {
+    return helper.dal.context.save(instanceArray.pop())
+    .then(function() {
+      if (instanceArray.length > 0) {
+        return saveInstanceArray(instanceArray);
```
Nice!

samuelg commented on test/context.js:
```diff
@@ -89,6 +90,58 @@ describe('context', function () {
+    }
+
+    var x;
+    for (x in result) {
```
Just use result.forEach(function(item) {}) instead of having to declare a var for x.

samuelg commented on test/context.js:
```diff
@@ -89,6 +90,58 @@ describe('context', function () {
+    var x;
+    for (x in result) {
+      var domain = result[x].domain;
+      var index = expectedItems.indexOf(domain);
```
Consider using expectedItems.filter().

samuelg commented on test/context.js:
```diff
@@ -89,6 +90,58 @@ describe('context', function () {
+    var instances = [];
+    var x;
+
+    for (x in items) {
```
Again, just use items.forEach().

samuelg commented on test/folder.js:
```diff
@@ -111,6 +111,59 @@ describe('folder', function () {
+    }
+
+    var x;
+    for (x in result) {
```
result.forEach()

samuelg commented on test/folder.js:
```diff
@@ -111,6 +111,59 @@ describe('folder', function () {
+      var item;
+      var y;
+
+      for (y in expectedItems) {
```
expectedItems.filter()

samuelg commented on test/mailbox.js:
```diff
@@ -130,6 +130,51 @@ describe('mailbox', function () {
+    }
+
+    var x;
+    for (x in result) {
```
result.forEach()

samuelg commented on lib/repositories/message.js:
```diff
@@ -394,20 +412,89 @@ function createApi(table, provider, sqlGenerator, dependencies) {
+                .then(function() {
+                  dependencies.logger.debug({messages: messages},
+                                            'Messages removed');
+                  return messages;
```
Why are messages being returned here? The messages won't be fully formed model objects since it doesn't use the common helper methods. The purpose of convertFromStorage is just to get the 'read' property to be converted to a boolean. Not sure this is necessary here. It's used in the message repository to only set the message as read once since it updates MWI and we don't want to duplicate that call.

samuelg replied:
Just remembered I suggested this. I think you mentioned that since the mailbox is being deleted, the MWI mailbox can just be delete via ARI by the caller. I think that would be best.

jrrose committed:
Address review feedback

samuelg commented:
Looks good to me. Since this has a lot of independent commits, we may want to look at squashing them and opening a new PR before merging in. I can walk you through that.

jrrose commented:
Squashed commits into #2
